### ISSUE TITLE
How to Identify Network Adapters in Powershell

### ISSUE THREAD

**Issue by dmidlo** [dmidlo/blog#4]
### Tldr:

To get your active network adapters:
```PowerShell
Get-NetAdapter -Physical | where status -eq "up"
```

#### Discussion

To get the physical media type of your adapters

```Powershell
Get-NetAdapter | select Name, PhysicalMediaType
```

One fine way to do it is with good'ol WMI using the `Get-WMIObject" cmdlet:

```Powershell
 Get-WmiObject win32_networkadapterconfiguration | Format-Table
```

maybe you'd like just the see the IP enabled adapters:

```Powershell
 Get-WmiObject win32_networkadapterconfiguration -Filter ‘ipenabled = “true”‘
```

You might be looking for the mac address, for that you'll use the `win32_networkadapter` namespace:

```PowerShell
Get-WmiObject win32_networkadapter | Format-Table
```

Powershell has a 'native' cmdlet that gives similar information

```PowerShell
Get-NetIPConfiguration | Format-Table
```

Use WMI to look up the connected adapters:

```PowerShell
Get-WmiObject win32_networkadapter -Filter "netconnectionstatus = 2" | ft
```

these are the possible status codes for the `netconnectionstatus` property
|val|def|
|-|-|
|0|Disconnected|
|1|Connecting|
|2|Connected|
|3|Disconnecting|
|4|Hardware not present|
|5|Hardware disabled|
|6|Hardware malfunction|
|7|Media disconnected|
|8|Authenticating|
|9|Authentication succeeded|
|10|Authentication failed|
|11|Invalid Address|
|12|Credentials required|

Using Powershell `Get-NetAdapter` to check connection status:

```PowerShell
Get-NetAdapter "Ethernet" | select Name, MediaConnection State
```

to get IP enabled adapters you can:

```Powershell
Get-NetIPInterface | ft
```

To disable adapter:

```Powershell
Disable-NetAdapter -Name "Ethernet"

# and to enable
Enable-NetAdapter -Name "Ethernet"
```

you can also use `Get-CimInstance` to get the same information from the `win32_networkadapter` namespace:

```PowerShell
Get-CimInstance -ClassName win32_networkadapter | ft
```

The `win32_networkadapter` has a property called `AdapterTypeId`, these are the possible values:

|val|def|
|-|-|
|0|Ethernet 802.3|
|1|Token Ring 802.5|
|2|Fiber Distributed Data Interface (FDDI)|
|3|Wide Area Network (WAN)|
|4|LocalTalk|
|5|Ethernet using DIX header format|
|6|ARCNET|
|7|ARCNET (878.2)|
|8|ATM|
|9|Wireless|
|10|Infrared Wireless|
|11|Bpc|
|12|CoWan|
|13|1394|

There are 41 properties on the WMI namespace:

```PowerShell
(Get-CimInstance -ClassName win32_networkadapter | Get-Member -Type Property).length
```

There are 103 properties on for the Powershell NetAdapter object

```PowerShell
(Get-NetAdapter | Get-Member -MemberType Property).length
```
To Look at all the properties in a compact view:

```PowerShell
$NetAdapterPropertyNames = Get-NetAdapter | Get-Member -MemberType Property | foreach {$Name = $_.Name; $Name}
$NetAdapterPropertyNames -join ", "
```

old-fashioned `netsh` works, but obviously only the text output to stdout would be pipelined:

```PowerShell
netsh interface ipv4 show interfaces
```

if you're using wireshark or otherwise using your interfaces to sniff on network traffic, you can find which interfaces are being used like this:

```PowerShell
Get-NetAdapter | ? PromiscuousMode -eq $true
```

To find out what services are running on reach adapter:

```PowerShell
Get-NetAdapterBinding
```
